tensorflow/contrib/opt/python/training/elastic_average_optimizer.py

```diff
@@ -0,0 +1,279 @@
+    """
+    self._worker_device = worker_device
+
+  def __call__(self, getter, name, *args, **kwargs):
+    if name.find('global_step') != -1:
+      return getter(name, *args, **kwargs)
+
+    kwargs['collections'] = [ops.GraphKeys.LOCAL_VARIABLES]
+    with ops.device(self._worker_device):
+      local_var = getter(name, *args, **kwargs)
+
+    if kwargs['trainable']:
+      global_center_variable = variables.Variable(
+        name='%s/%s' %
+             ('global_center_variable',
+              name),
+        initial_value=local_var.initialized_value(),
+        trainable=False,
+        collections=[
+          ops.GraphKeys.GLOBAL_VARIABLES,
+          'global_center_variable'])
+
+      with ops.device(self._worker_device):
+        local_center_variable = variables.Variable(
+          name='%s/%s' % ('local_center_variable', name),
+          initial_value=local_var.initialized_value(),
+          trainable=False,
+          collections=[ops.GraphKeys.LOCAL_VARIABLES,
+                       'local_center_variable'])
+    return local_var
+
+
+class ElasticAverageOptimizer(optimizer.Optimizer):
+  """Wrapper optimizer that implements the Elastic Average SGD algorithm.
+  This is an async optimizer. During the training, Each worker will update
+  the local variables and maintains its own local_step, which starts from 0
+  and is incremented by 1 after each update of local variables. Whenever
+  the communication period divides the local step, the worker requests
+  the current global center variables and then computed the elastic difference
+  between global center variables and local variables. The elastic difference
+  then be used to update both local variables and global variables.
+  """
+
+  def __init__(
+      self,
+      opt,
+      num_worker,
+      communication_period=10,
+      moving_rate=None,
+      rho=None,
+      use_locking=True,
+      name="ElasticAverageOptimizer"):
+    """Construct a new gradient descent optimizer.
+
+    Args:
+      opt: The actual optimizer that will be used to update local variables.
+        Must be one of the Optimizer classes.
+      communication_period: An int point value to controls the frequency
+        of the communication between every worker and the ps.
+      num_worker: The number of workers
+      moving_rate: A floating point value to control the elastic difference.
+      rho: the amount of exploration we allow ine the model. The default
+        value is moving_rate/learning_rate
+      use_locking: If True use locks for update operations.
+      name: Optional name prefix for the operations created when applying
+        gradients. Defaults to "GradientDescent".
+    """
+    super(ElasticAverageOptimizer, self).__init__(use_locking, name)
+    self._opt = opt
+    self._num_worker = num_worker
+    self._period = communication_period
+
+    if moving_rate is None:
+      self._moving_rate = 0.9 / communication_period / num_worker
```

Comment:
Why 0.9 here? If it is reasonable, you'd better define a named constant instead of `0.9`.